### ISSUE TITLE
[DISCUSS] Apache PouchDB

### ISSUE THREAD

**Issue by janl** [apache/pouchdb#8999]
*This issue proposes to move the PouchDB project to the Apache Software Foundation. It outlines the pros and cons, alternatives and next concrete steps.*

*About the Author: Hi, I’m Jan Lehnardt, I’m the Project Management Chair of Apache *CouchDB* and its longest standing contributor. I currently informally also have taken over managerial duties for PouchDB.*

# Introduction

PouchDB has been chiefly run by Dale Harvey and Nolan Lawson in the past. Both have since stepped away from the project and a mixture of folks from the PouchDB user community have stepped up to continue improving and maintaining PouchDB. So far so good, Open Source at work.

PouchDB, while not immensely popular, is being used in a wide variety of FOSS projects and proprietary products, including a number of mission-critical, humanitarian and medical environments where no other software solution would suffice. PouchDB is important.

Project management and progress as well as decision making has been ad-hoc since the principal developers have left. This has worked without too many issues so far, but I believe PouchDB would serve its users better and be a more long-term reliable choice for new developments if its project management could be formalised.

Now, nothing stops us from writing up processes and adhering to them, but I also think it would strengthen PouchDB’s position if it joined an Open Source Foundation. The two immediate choices for this are the [OpenJS Foundation](https://openjsf.org/) and the [Apache Software Foundation](https://www.apache.org/), the former because it focuses on JavaScript projects and the latter because it is already the home of Apache CouchDB, PouchDB’s sister project.

In preliminary talks with the currently active project folks, we are aiming for joining the ASF to be close to CouchDB.

# Foundations

## Benefits of Joining a Foundation

An Open Source Foundation helps with a number of things smaller-scale projects do not have to deal with, but that when handled correctly and with experience, elevate the project in question. This include, but are not exclusive to:

* Help with any legal issues by having lawyers on stand-by.
* Help with any trademark issues that might arise.
    * *Author’s note: this alone is invaluable based on my experience with CouchDB.*
* Help with ready-made processes for project governance:
    * Who can decide what? How do we make sure one person doesn’t make all the decisions or vetoes important changes? How do we make sure the project does not stall out if a key person leaves? How do we make sure we have a steady stream of new contributors? How do we make sure the project is viable long-term? How do we make sure no one company with all the money dominates the project?
* Help with ensuring best practices around software security and release engineering.
* Industry recognition: joining a foundation can give a project an additional sheen of legitimacy.

## Why the ASF?

Apache CouchDB, PouchDB’s sister project, is already with the ASF. Having PouchDB join makes that relationship even more clear. It also allows folks betting on CouchDB and PouchDB to have the same rules of governance apply to both projects. Finally, we can benefit from some of the current PouchDB project folks’ experience with the ASF.

## ASF Caveats

*Author’s note: I have been with CouchDB since before it joined the ASF and I can speak to being part of an ASF project for 15 years now. I’ve learned the ins and outs the hard way and I’m happy to distil my knowledge for the benefit of this proposal.*

The ASF’s *chief goal is longevity of software projects*. It cares very little about what technical things are going on inside a project, as long as its best practices for long-term survival of open source projects are followed.

Many of the caveats I raise here are to be evaluated relative to this primary goal. The ASF is extremely successful with this strategy. The foundation itself is over 25 years old and hosts some of the biggest open source projects there are.

The caveats:

* The ASF moves slowly. A new project joining will always have its own ideas of how to do certain things. And while it is within each project’s power to help shape the ASF in the way it needs it, it might take more time and commitment than you are willing to give at any one point. Changes come slowly to the ASF, but eventually, the right things tend to happen.
* Email. The ASF uses email as its predominant communication medium. Every significant project decision has to be made through one of its mailing lists. This might feel unfamiliar with folks who have grown up using GitHub for everything, but do expect to read and write emails a lot. You will also learn how to filter emails. Luckily, most development-related work and discussion can continue to happen on GitHub because the ASF mirrors these discussions to its own infrastructure’s mailing lists. You see, longevity means: how can we read up on project decision making from the past even if GitHub is no longer around? The ASF has longevity at its core for everything it does.
* Processes: the ASF is a very lean organisation, projects report to a board, but the board does not take any direct project action as long as a handful of guidelines are followed. That means on the one hand that a project is relatively free to define how it works, but on the other hand, it means that it has to formalise how it works a little more than it probably does outside of the ASF. For PouchDB, we can steal a few best-practices and [bylaws documents](https://couchdb.apache.org/bylaws.html) from CouchDB to get started. I think formalising how PouchDB is developed is a benefit, even if it might seem daunting at first.
* The Name: The ASF has been called out repeatedly for appropriating native american naming and imagery. While no public action has been taken to remedy this (as far as the author knows at the time of writing), this is something that the ASF is actively considering. At that point, most likely, the Apache Software Foundation will simply be known as The ASF eventually. If we are ever asked about this, we can say that we are aware of the issue and support any changes.

# Next Steps

Okay what now?

1. Collect the initial team. Anyone who has a current or past stake in PouchDB and who I think are beneficial to the future of PouchDB can join. Once we have the team together, we can take the next steps of actually starting the process of joining the ASF. A shortlist of existing contributors have seen a draft version of this document and have intimated that they are interested in joining.
    1. A secondary consideration for the selection was breadth of organisations. At the moment, folks who are employed at Medic Mobile and Neighbourhoodie make out the majority of folks. My goal here is to find a set of folks that ensure that neither organisation has a majority stake in contributors, and that we have a sizable group of non-affiliated folks to keep the balance.
2. The proposed team (if you are on the list below and no longer want to join, or if you are not on this list and do want to join, please leave a comment):
    1. Alba Herrerías Ramírez @AlbaHerrerias (Neighbourhoodie)
        1. 8.0.0 release manager & feature developer
    1. Alex Anderson @alxndrsn (independent, formerly Medic)
        1. current core contributor
    1. Diana Thayer @garbados (independent)
        1. Author of many PouchDB plugins, like crypto-pouch
    1. Diana Barsan @dianabarsan (Medic)
        1. current contributor
    1. Gareth Bowen @garethbowen (Medic)
        1. current core contributor
    1. Jan Lehnardt @janl (Neighbourhoodie)
        1. current core contributor
    1. James Coglan @jcoglan (Neighbourhoodie)
        1. 8.0.0 release manager and feature developer
    1. Johannes Schmidt @jo (Mozilla / independent)
        1. long-time contributor
    1. Steven-John Lange @SourceR85 (HZData GmbH / independent)
        1. current core contributor
    1. Tally:
        1. Medic: 2
        2. Neighbourhoodie: 3
        3. Independent: 4
3. The ASF Incubator
    1. The process  of joining any foundation usually starts with a trial period. With the ASF, this is called the Incubation. The ASF Incubator is itself an ASF project and its job is to bring new projects into the ASF, make them learn [the way of the ASF](https://www.apache.org/theapacheway/) and help them graduate to become a *Top Level ASF Project.* The process typically takes 6–12 months, depending on how quickly we hit all the required milestones. The current process is [well documented in The Apache Incubator Cookbook](https://incubator.apache.org/cookbook/).
4. Apache PouchDB
    1. Once we have demonstrated we have understood and followed all the processes, we can ask to graduate from the incubator and become *Apache PouchDB*.

There are a number of things that will have to change or be introduced to PouchDB, to make sure we can join and graduate from the ASF Incubator. Everything necessary is outlined or linked from the Incubator Cookbook, but from my experience with the ASF, here are the top level items:

1. We can continue using GitHub. There is just a small setup step required to make things work with the ASF infrastructure.
2. We all have to sign a [contributor licence agreement](https://apache.org/licenses/contributor-agreements.html) (CLA, [FAQ here](https://apache.org/licenses/cla-faq.html)) that gives the ASF the legal right to all the code we write, so it can protect it for us. This is a small legal document that is not very controversial, but do let me know if you have any questions.
3. We will have to start using email for a few types of communication. The ASF will create mailing lists for us that we can use.
4. We will have to absolutely make sure PouchDB doesn’t have any dependencies that have software licences that are [not compatible](https://apache.org/legal/resolved.html) with the ASF. Most prominently this is the GPL-family of licences. I think we are okay overall, the [licensee](https://github.com/jslicense/licensee.js) tool can help us formalise this.
5. We might have to fork a couple of GitHub Actions to live on ASF repositories for our CI to continue running.
6. The pouchdb/pouchdb repo will be migrated to apache/pouchdb (whether we fork or move is up to us to decide later). We can also move any pouchdb/* subprojects to apache/pouchdb-*
7. The main project domain is going to be [https://pouchdb.apache.org](https://pouchdb.apache.org)
    1. We can keep pouchdb.com as a redirect there and it is sometimes useful to have auxiliary subdomains available.
    2. The ASF does not take on and manage what it calls “vanity domains” for projects, which does include pouchdb.com, but I’m happy to put the domain under Neighbourhoodie control in escrow for the project. This ensures continuity beyond one person and we don’t lose any links. We are already doing this for CouchDB vanity domains.
8. We will have to follow the requirements of the official [ASF software release policy](https://www.apache.org/legal/release-policy.html). While this might feel like more ceremony than is useful at first, the steps in there are again valuable for making sure we release long-term reliable software.
9. TBD: I’ll fill in these ase they come up

# FAQ

I’m going to evolve this as questions come in.

## What is expected of me?

You are reading this because I think you have PouchDB’s best interests in mind. I’d like your opinion on all of the above as well as all the steps through the ASF to become a top level project. And after that help shape PouchDB’s future.

This might all seem very abstract. It will mostly be first learning about how the ASF works and then discussing with everybody else on this team on how to apply the ASF policies to how we want PouchDB to work. This will be a one-time effort to get through the Incubator.

From there on out, you will have the responsibility to help PouchDB be successful based on all the decisions we worked out.

In practice, this means reading and writing a bunch of emails, more in the first 6 months, less later.

Most importantly, in accordance with the ASF credo *community over code*, you are under no real obligation at all. If you need to step away from the project temporarily or permanently, that is absolutely no problem.

If you just want to focus on a certain aspect of this process (say how to get new contributors into the project, or just the release procedure), then that is also just fine.

## How much of a time commitment is this?

My current expectation is 1–2 hours per week for the next ~12 months, less afterwards.

## Where can I learn more about the ASF?

Start here: [https://www.apache.org/foundation/how-it-works/](https://www.apache.org/foundation/how-it-works/)

—

I’m excited about the future of PouchDB, what do you all think?

**Comment by garethbowen** [comment]
Pouch is a critical component of some important work, so +1 for longevity. Let's do it.

**Comment by dianabarsan** [comment]
I'm all for this. Thank you for the initiative!

**Comment by SourceR85** [comment]
Thank you for this great explanation 👍

Joining the ASF also seems to me to be the right way to go.

**Comment by nickva** [comment]
+1

**Comment by jo** [comment]
:100: no-brainer

**Comment by ChiragMoradiya** [comment]
+1

**Comment by lucidNTR** [comment]
Thanks for the initiative, from a user/advocate perspective this would be fantastic, as long as the team can manage the overhead, i hope a big part will be automatable or one time changes.

**Comment by garbados** [comment]
Hell yeah. Let's roll!

**Comment by garrensmith** [comment]
Awesome. +1. Its great to see PouchDB still being worked on and maintained.

**Comment by sfuentes** [comment]
As a longtime lurker and advocate of pouch/couch solutions, I would be happy to know that the Project is under a Foundation Umbrella. +1

**Comment by janl** [comment]
Thank you all for the comments and support!

As per [the incubation process](https://incubator.apache.org/cookbook/#does_our_project_fit_the_apache_incubator) our next step is “Find a champion and incubation mentors and discuss and prepare an incubation proposal” — I’ll draft communication with the incubator shortly.

I’m also preparing a first draft of the Incubation Proposal for the next+1 step. I’ll open a separate PR for that.

We’ll keep this issue open for status updates.

Thanks again 🙏

**Comment by janl** [comment]
> I’ve just sent the solicitation email: 

We gained two Mentors and I’ll be accent as Champion, if the Incubator PMC will have me. Meanwhile, I’ll continue on our application doc: https://github.com/pouchdb/pouchdb/pull/9060

**Comment by janl** [comment]
And we’re off: https://lists.apache.org/thread/j6j6x1t6ppdqttth2cfnr8rb7v1n4yrv

**Comment by janl** [comment]
Hey all, I have called an official ASF VOTE™ on the addition of PouchDB to the ASF: https://lists.apache.org/thread/z6slrvwvmvh45n8tf34qsqr9m28rokdq

It would be great to show your support by voting +1 on the proposal on the mailing list. To subscribe, send an email to general-subscribe@incubator.apache.org and respond to the confirmation email you received.

Then to vote on the thread, send an email to general@incubator.apache.org with these additions:

1. Add this header: `In-Reply-To: <8091F358-998F-40BF-ABB8-8C8CD634D95D@apache.org>`
2. Use this subject line: `Re: [VOTE] Accept PouchDB into the Apache Incubator`
3. In the body, you can vote by writing `+1` plus any other supportive text you like.

Your votes will not be “binding” (they don’t affect the final decision), but showing up as a group will show good dynamics and our commitment to the project.

Thanks!

**Comment by janl** [comment]
Heya, I wanted to keep you posted on progress.

As linked before, we have been formally accepted into the ASF incubator.

Before we can do any actual work, some infrastructure pieces need to be put in place by the ASF Infra team. That’s currently ongoing. After that, we can create further project infrastructure, like accounts, repos, mailing lists and so on.

I’ll keep folks posted as things move along.

**Comment by pjfanning** [comment]
Would it be possible to get a fuller status update?
https://cwiki.apache.org/confluence/display/INCUBATOR/September2025#pouchdb is fine for the ASF Board
There isn't really enough activity on the dev@pouchdb.apache.org to discuss progress there yet.

Would it be possible to enumerate the blockers and what else is ongoing about getting the source transferred to the apache git namespace?

**Comment by alxndrsn** [comment]
> Would it be possible to enumerate the blockers and what else is ongoing about getting the source transferred to the apache git namespace?

Per https://lists.apache.org/thread/rxt5xnrlsofnkbn6dw3o6p619r6lbkm6:

> 1. Audit Cryptography https://infra.apache.org/crypto.html
> 2. Initial Clean Up https://incubator.apache.org/guides/transitioning_asf.html
> 3. Ask ASK Infra to transfer our repo (I’ll volunteer for this)

There has been some discussion of **1** at https://lists.apache.org/thread/6dnwkyyt24pgqzgp31fk6jyymdv6hc7t, and in #incubator in the ASF slack.  There is also some discussion in the couchdb slack.

**1** is blocking **2** and **3**.

**Comment by pjfanning** [comment]
@alxndrsn the answer to 1 appears to be that noone in ASF is concerned - can you correct me if anyone in ASF has explicitly blocked you on 1?

**Comment by alxndrsn** [comment]
> [@alxndrsn](https://github.com/alxndrsn) the answer to 1 appears to be that noone in ASF is concerned - can you correct me if anyone in ASF has explicitly blocked you on 1?

Sorry if I implied that - I was hoping for some discussion between the other team members.  Personally I'm nervous about rewriting git history, but it also seems very practical.

**Comment by pjfanning** [comment]
Noone in ASF has asked you to rewrite the git history.
The ASF is most concerned by releases endorsed by the ASF (ie reviewed and voted on using the dev@pouchdb.apache.org list and published to the ASF dist.apache.org svn repo). The key release artifact is a source tarball that is a point in time snapshot of the code.
I cannot see why this is blocking progress.
If someone asks for some stuff to be removed from git history which appears unlikely, that can happen even after the git repo is moved to Apache namespace.

**Comment by alxndrsn** [comment]
> I cannot see why this is blocking progress.

So can we move this repo as-is to the apache GitHub organisation?

**Comment by pjfanning** [comment]
> > I cannot see why this is blocking progress.
> 
> So can we move this repo as-is to the apache GitHub organisation?

I don't see why not. The podling and its resources are not expected to be perfect from day one. The aim is to improve over time.

**Comment by janl** [comment]
With the repo having moved, I think we can close this now. Thanks all for playing!
